### Задание 1
Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [57]:
import pandas as pd
import numpy as np

from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

In [58]:
from sklearn.linear_model import LinearRegression

boston = load_boston()

data = boston["data"]
feature_names = boston["feature_names"]
X = pd.DataFrame(data, columns=feature_names)

target = boston["target"]
y = pd.DataFrame(target, columns=["price"])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

regr_lr = LinearRegression()
regr_lr.fit(X_train, y_train)

y_pred = regr_lr.predict(X_test)

print(f'Mean squared error (MSE): {mean_squared_error(y_test, y_pred):.2f}')
print(f'Coefficient of determination (R2): {r2_score(y_test, y_pred):.2f}')

Mean squared error (MSE): 21.52
Coefficient of determination (R2): 0.71


### Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [59]:
from sklearn.ensemble import RandomForestRegressor

regr_rfr = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)
regr_rfr.fit(X_train, y_train.values[:,0])

y_pred = regr_rfr.predict(X_test)

print(f'Mean squared error (MSE): {mean_squared_error(y_test, y_pred):.2f}')
print(f'Coefficient of determination (R2): {r2_score(y_test, y_pred):.2f}')

Mean squared error (MSE): 9.33
Coefficient of determination (R2): 0.87


#### Модель с RandomForestRegressor работает лучше, MSE=9.33 против 21.52. R2 также лучше у второй модели.

### Задание 3
Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.
С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.

In [60]:
feat_arr = regr_rfr.feature_importances_
feat_arr.sum()

1.0

In [61]:
feat_arr_imp = sorted(range(len(feat_arr)), key=lambda x: feat_arr[x])[-2:]
for i in feat_arr_imp:
    print(f'{list(X_test)[i]}={feat_arr[i]:.2f}')

RM=0.40
LSTAT=0.42
